### PR TITLE
Switch to sabre/vobject and add support VTIMEZONE for ICS link

### LINKED ISSUE #73: Missing VTIMEZONE component in ics file
Hello

When importing an ics file generated by spatie/calendar-links in Outlook 365, the datetime may be in the wrong timezone.
See this thread for a true life story: https://techcommunity.microsoft.com/t5/Office-365-Calendar/Calendar-incorrectly-reading-ics-file-time-and-timezone-are/m-p/715011/highlight/true#M327

According to the Microsoft tech guy, and after reading the [vcalendar rfc](https://tools.ietf.org/html/rfc5545) , it is clear that if a date is given with a timezone, a VTIMEZONE component must be present in the ics file.

### PR DESCRIPTION
**Pull request by dogawaf**
This PR introduces the usage of [sabre/vobject](http://sabre.io/vobject/) for generating ics file.
sabre/vobject is a pretty stable and robust library.

The escaping of strings and the date and timezone handling is delegated to the library.

This PR also adds a VTIMEZONE component when the event features dates that are not on UTC.
VTIMEZONE is mandatory for any TZID parameter on dates of VEVENT (according rfc).
This fixes an issue when importing an ics on Outlook, which cannot handle properly timezone if the ics file have no VTIMEZONE component.

Fixes #73

### CODE REVIEW

**Comment by tusharvikky** [comment]
Hello Ange,

Any update on the merge?

Regards

**Comment by atymic** [comment]
cc @freekmurze 

Any chance of a review here? Thanks :)

**Comment by atymic** [comment]
@freekmurze

**Comment by atymic** [comment]
cc @freekmurze

File: src/Generators/Ics.php [hunk 1]
```diff
@@ -55,4 +66,107 @@ protected function generateEventUid(Link $link): string
...
+        $vTimeZone->TZID = $timeZone->getName();
+
+        $std = null;
+        $dst = null;
+        foreach ($transitions as $i => $trans) {
+            $component = null;
+
+            if ($i === 0) {
+                // remember the offset for the next TZOFFSETFROM value
+                $tzfrom = $trans['offset'] / 3600;
```
Comment (lptn) [inline src/Generators/Ics.php:124]: there is a missing `continue` statement here: without it `tzfrom` and `offset` is the same for the first VTIMEZONE (and thus TZOFFSETFROM and TZOFFSETTO also the same, what's wrong in most cases)

File: src/Generators/Ics.php [hunk 2]
```diff
@@ -55,4 +66,107 @@ protected function generateEventUid(Link $link): string
...
+        if (! $from) {
+            $from = time();
+        }
+        if (! $to) {
+            $to = $from;
+        }
+
+        // get all transitions for one year back/ahead
+        $year = 86400 * 360;
+        $transitions = $timeZone->getTransitions($from - $year, $to + $year);
```
Comment (lptn) [inline src/Generators/Ics.php:112]: for some  timezones without DST it's not enough. E.g. latest 4 transactions for `Europe/Minsk` tz are:
```json
[
    {"ts":1269734400,"time":"2010-03-28T00:00:00+0000","offset":10800,"isdst":true,"abbr":"EEST"},
    {"ts":1288483200,"time":"2010-10-31T00:00:00+0000","offset":7200,"isdst":false,"abbr":"EET"},
    {"ts":1301184000,"time":"2011-03-27T00:00:00+0000","offset":10800,"isdst":false,"abbr":"+03"},
    {"ts":2147483647,"time":"2038-01-19T03:14:07+0000","offset":10800,"isdst":false,"abbr":"+03"},
]
```
If you want to generate VTIMEZONE for e.g. 2014-03-01 - it will be empty

File: src/Generators/Ics.php [hunk 3]
```diff
@@ -55,4 +66,107 @@ protected function generateEventUid(Link $link): string
...
+                $dst = $vcalendar->createComponent('DAYLIGHT');
+                $component = $dst;
+            } // standard time definition
+            else {
+                $t_std = $trans['ts'];
+                $std = $vcalendar->createComponent('STANDARD');
+                $component = $std;
+            }
+
+            if ($component) {
```
Comment (lptn) [inline src/Generators/Ics.php:139]: this condition is extra: `$component` is always a `Component` instance (see if-else above)

File: src/Generators/Ics.php [hunk 4]
```diff
@@ -55,4 +66,107 @@ protected function generateEventUid(Link $link): string
...
+                $offset = $trans['offset'] / 3600;
+
+                $component->DTSTART = $dt->format('Ymd\THis');
+                $component->TZOFFSETFROM = sprintf('%s%02d%02d', $tzfrom >= 0 ? '+' : '', floor($tzfrom),
+                    ($tzfrom - floor($tzfrom)) * 60);
+                $component->TZOFFSETTO = sprintf('%s%02d%02d', $offset >= 0 ? '+' : '', floor($offset),
+                    ($offset - floor($offset)) * 60);
+
+                // add abbreviated timezone name if available
+                if (! empty($trans['abbr'])) {
```
Comment (lptn) [inline src/Generators/Ics.php:150]: Do you know any cases when abbr is empty string or null or false?

File: src/Generators/Ics.php [hunk 5]
```diff
@@ -55,4 +66,107 @@ protected function generateEventUid(Link $link): string
...
+                    $component->TZNAME = $trans['abbr'];
+                }
+
+                $tzfrom = $offset;
+                $vTimeZone->add($component);
+            }
+
+            // we covered the entire date range
+            if ($std && $dst && min($t_std, $t_dst) < $from && max($t_std, $t_dst) > $to) {
```
Comment (lptn) [inline src/Generators/Ics.php:159]: What is the reason on why do we need both `$std` and `$dst`?

File: src/Generators/Ics.php [hunk 6]
```diff
@@ -55,4 +66,107 @@ protected function generateEventUid(Link $link): string
...
+        if (! $to) {
+            $to = $from;
+        }
+
+        // get all transitions for one year back/ahead
+        $year = 86400 * 360;
+        $transitions = $timeZone->getTransitions($from - $year, $to + $year);
+
+        $vTimeZone = $vcalendar->createComponent('VTIMEZONE');
+        $vTimeZone->TZID = $timeZone->getName();
```
Comment (lptn) [inline src/Generators/Ics.php:115]: How about using `$component->add('TZID', $timeZone->getName())` syntax for the consistency with the code from above (e.g. `$vevent->add('DESCRIPTION', $link->description);`)?

**Review by lptn — CHANGES_REQUESTED** [review]
@dogawaf thanks a lot for your amazing work!

Sorry for such long delay. It caused main by 2 reasons:
 1. This code is difficult to test (reason: we need to know ics specs very well + all edge cases) 
 2. This PR uses a 3rd party package that may slow down our package

We can live with 2), but in order to solve 1) we had few options: ship a solution that works in 80% cases or test it, understand the code and find pitfalls.

I've spent about 4 hours reading specs and this code, trying to understand how it works, how we can improve it. Based on my research, I have some questions, please check them. I hope, together we can improve and merge this PR.

And again, thanks a lot for your work!

**Comment by philwolstenholme** [comment]
I still think this would be a useful change, and it seems a shame to close something that has had a code review and is waiting on some changes

**Comment by dogawaf** [comment]
I created this PR and use it on a specific project, and did not have any issue with it since. But merging it as is would not be a very good idea. I agree with @lptn on pretty much all the remarks.
I did not have the resource since to fix the remarks and add some tests (though sabre/vtimezone is already pretty well tested), but who knows... 
If somebody wan't to complete this PR and start again the review process, got for it.
Or, you can of course apply the patch for your specific need (like I did).
Cheers.

**Comment by abferris** [comment]
Hi there! I'm working on trying to wrap some of these up as a client has had issues. I'm not sure if all of this still applies, but I'm doing my best to work out exactly how all of it works. I'll be putting in comments in response to the above changes.

**Comment by bboro** [comment]
Hi @lptn @abferris

I've worked on some of the PR review suggestions, and also rerolled this against latest master (212a8c1). As I didn't have permission to push to @dogawaf's fork, I've created a PR here from my own fork to this PR: https://github.com/dogawaf/calendar-links/pull/3

I understand with all the changes coming in from master, it might be difficult to see what's changed. So here's a PR from the same branch against 212a8c1: https://github.com/bboro/calendar-links/pull/2/files

It would be great to move this forward and merged! Hope this helps.

**Comment by mitcks** [comment]
@bboro have you had any updates on this fix getting merged in?

**Comment by bboro** [comment]
@mitcks not yet
